gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java

```diff
@@ -1470,94 +1480,195 @@ public default GraphTraversal<S, String> concat(final String... concatStrings) {
     /**
      * Returns the value of incoming traverser as strings. Null values are returned as a string value "null".
      *
-     * @return the traversal with an appended {@link AsStringStep}.
+     * @return the traversal with an appended {@link AsStringGlobalStep}.
      * @see <a href="http://tinkerpop.apache.org/docs/${project.version}/reference/#asString-step" target="_blank">Reference Documentation - AsString Step</a>
      * @since 3.7.1
      */
     public default GraphTraversal<S, String> asString() {
         this.asAdmin().getBytecode().addStep(Symbols.asString);
-        return this.asAdmin().addStep(new AsStringStep<>(this.asAdmin()));
+        return this.asAdmin().addStep(new AsStringGlobalStep<>(this.asAdmin()));
+    }
+
+    /**
+     * Returns the value of incoming traverser as strings. Null values are returned as a string value "null".
+     *
+     * @param scope local will operate on individual strings within incoming lists, global will operate on current traversal as a single object.
+     * @return the traversal with an appended {@link AsStringGlobalStep} or {@link AsStringLocalStep} depending on the {@link Scope}.
+     * @see <a href="http://tinkerpop.apache.org/docs/${project.version}/reference/#asString-step" target="_blank">Reference Documentation - AsString Step</a>
+     * @since 3.7.1
+     */
+    public default GraphTraversal<S, E> asString(final Scope scope) {
```

Comment:
I'm concerned about the use of type inference here. Given that the return value could be a String or a List and an unchecked cast is used to return E. Can you confirm that you've tested to make sure this doesn't cause casting errors at runtime?

Code like:
`Integer shouldntCast = g.V().asString(Scope.global).next()` should cause a compilation issue and not a runtime cast exception.

Reply:
The cast compilation warning occurs when the assignment type doesn't match the `S` type. Unfortunately I'm not sure if there is a better way to work around the casting with returning String or List, do you have any suggestions?

Reply:
Update to use `E2`